File: src/mk1/PythonPackage/src/tamaf/fipacontractnetresponder.py
from typing import TYPE_CHECKING
from .fsmbehavior import FSMBehavior
from .simplebehavior import SimpleBehavior
from .oneshotbehavior import OneShotBehavior
from .aclmessagetemplate import ACLMessageTemplate
from .performative import Performative

if TYPE_CHECKING:
    from .agent import Agent
    from .aclmessage import ACLMessage

class FIPAContractNetResponder(FSMBehavior):
    
    def __init__(self, agent: 'Agent', cfpTemplate: 'ACLMessageTemplate'):
        super().__init__(agent)
        self.cfpTemplate = cfpTemplate

        class WaitCfpBehavior(OneShotBehavior):
            def __init__(self, fsm: 'FIPAContractNetResponder'):
                super().__init__(fsm.agent)
                self.fsm = fsm

            def Action(self):
                cfpMsg = self.Receive(self.fsm.cfpTemplate)
                self.fsm.sharedData['cfp'] = cfpMsg
                
                self.fsm.sharedData['reply_template'] = ACLMessageTemplate(
                    conversationID=cfpMsg.conversationid
                )

            def OnEnd(self):
                return 0

        class PrepareResponseBehavior(SimpleBehavior):
            def __init__(self, fsm: 'FIPAContractNetResponder'):
                super().__init__(fsm.agent)
                self.fsm = fsm
                self.transitionCode: int
                self.is_done: bool

            def OnStart(self):
                super().OnStart()
                self.is_done = False
                self.transitionCode = -1

            def Action(self):
                cfpMsg = self.fsm.sharedData['cfp']
                reply = self.fsm.handleCfp(cfpMsg)
                if reply is None:
                    return

                if getattr(cfpMsg, 'conversationid', None):
                    reply.conversationid = cfpMsg.conversationid
                
                self.fsm.sharedData['propose_msg'] = reply
                self.agent.Send(reply)
                
                if reply.performative == Performative.PROPOSE:
                    self.transitionCode = 0 
                else:
                    self.transitionCode = 1 
                self.is_done = True

            def Done(self):
                return self.is_done

            def OnEnd(self):
                return self.transitionCode

        class WaitAcceptRejectBehavior(SimpleBehavior):
            def __init__(self, fsm: 'FIPAContractNetResponder'):
                super().__init__(fsm.agent)
                self.fsm = fsm
                self.transitionCode: int
                self.is_done: bool
            
            def OnStart(self):
                super().OnStart()
                self.is_done = False
                self.transitionCode = -1

            def Action(self):
                reply = self.Receive(self.fsm.sharedData['reply_template'])
                if reply is None:
                    return

                cfpMsg = self.fsm.sharedData['cfp']
                proposeMsg = self.fsm.sharedData['propose_msg']

                if reply.performative == Performative.ACCEPT_PROPOSAL:
                    self.fsm.sharedData['accept_msg'] = reply
                    self.transitionCode = 0 
                else:
                    res = self.fsm.handleRejectProposal(cfpMsg, proposeMsg, reply)
                    if res is None or res is False:
                        return
                    self.transitionCode = 1 
                self.is_done = True

            def Done(self):
                return self.is_done

            def OnEnd(self):
                return self.transitionCode

        class ExecuteTaskBehavior(SimpleBehavior):
            def __init__(self, fsm: 'FIPAContractNetResponder'):
                super().__init__(fsm.agent)
                self.fsm = fsm
                self.is_done: bool

            def OnStart(self):
                super().OnStart()
                self.is_done = False

            def Action(self):
                cfpMsg = self.fsm.sharedData['cfp']
                proposeMsg = self.fsm.sharedData['propose_msg']
                acceptMsg = self.fsm.sharedData['accept_msg']
                resultMsg = self.fsm.handleAcceptProposal(cfpMsg, proposeMsg, acceptMsg)
                if resultMsg is None:
                    return
                
                if getattr(cfpMsg, 'conversationid', None):
                    resultMsg.conversationid = cfpMsg.conversationid
                self.agent.Send(resultMsg)
                self.is_done = True

            def Done(self):
                return self.is_done

            def OnEnd(self):
                return 0 

        self.AddInitialState(WaitCfpBehavior(self), "waitCfp")
        self.AddState(PrepareResponseBehavior(self), "prepareResponse")
        self.AddState(WaitAcceptRejectBehavior(self), "waitAcceptReject")
        self.AddState(ExecuteTaskBehavior(self), "executeTask")

        self.AddTransition("waitCfp", "prepareResponse", 0)
        self.AddTransition("prepareResponse", "waitAcceptReject", 0) 
        self.AddTransition("prepareResponse", "waitCfp", 1)           
        self.AddTransition("waitAcceptReject", "executeTask", 0)     
        self.AddTransition("waitAcceptReject", "waitCfp", 1)         
        self.AddTransition("executeTask", "waitCfp", 0)               

    def handleCfp(self, cfpMessage: 'ACLMessage') -> 'ACLMessage':
        pass
    def handleAcceptProposal(self, cfpMessage: 'ACLMessage', proposeMessage: 'ACLMessage', acceptMessage: 'ACLMessage') -> 'ACLMessage':
        pass
    def handleRejectProposal(self, cfpMessage: 'ACLMessage', proposeMessage: 'ACLMessage', rejectMessage: 'ACLMessage') -> bool:
        return True

File: src/mk2/JavaPackage/src/main/java/tamaf/FIPAContractNetResponder.java
package tamaf;

public class FIPAContractNetResponder extends FSMBehavior {
    protected ACLMessageTemplate cfpTemplate;

    public FIPAContractNetResponder(Agent agent, ACLMessageTemplate cfpTemplate) {
        super(agent);
        this.cfpTemplate = cfpTemplate;

        addInitialState(new WaitCfpBehavior(this), "waitCfp");
        addState(new PrepareResponseBehavior(this), "prepareResponse");
        addState(new WaitAcceptRejectBehavior(this), "waitAcceptReject");
        addState(new ExecuteTaskBehavior(this), "executeTask");

        addTransition("waitCfp", "prepareResponse", 0);
        addTransition("prepareResponse", "waitAcceptReject", 0); // Bid sent
        addTransition("prepareResponse", "waitCfp", 1);           // Passed
        addTransition("waitAcceptReject", "executeTask", 0);     // Won auction
        addTransition("waitAcceptReject", "waitCfp", 1);         // Lost auction
        addTransition("executeTask", "waitCfp", 0);               // Task done
    }

    public ACLMessage handleCfp(ACLMessage cfp) { return null; }
    public ACLMessage handleAcceptProposal(ACLMessage cfp, ACLMessage propose, ACLMessage accept) { return null; }
    public Boolean handleRejectProposal(ACLMessage cfp, ACLMessage propose, ACLMessage reject) { return true; }

    private class WaitCfpBehavior extends OneShotBehavior {
        FIPAContractNetResponder fsm;
        WaitCfpBehavior(FIPAContractNetResponder fsm) { super(fsm.getAgent()); this.fsm = fsm; }
        @Override
        public void action() {
            ACLMessage cfpMsg = receive(fsm.cfpTemplate);
            if (cfpMsg == null) return;

            fsm.sharedData.put("cfp", cfpMsg);
            ACLMessageTemplate replyTemplate = new ACLMessageTemplate();
            replyTemplate.setConversationId(cfpMsg.getConversationId());
            fsm.sharedData.put("reply_template", replyTemplate);
        }
    }

    private class PrepareResponseBehavior extends SimpleBehavior {
        FIPAContractNetResponder fsm;
        int transitionCode;
        boolean isDone;

        @Override
        public void onStart() {
            super.onStart();
            this.transitionCode = -1;
            this.isDone = false;
        }

        PrepareResponseBehavior(FIPAContractNetResponder fsm) { super(fsm.getAgent()); this.fsm = fsm; }
        @Override
        public void action() {
            ACLMessage cfpMsg = (ACLMessage) fsm.sharedData.get("cfp");
            ACLMessage reply = fsm.handleCfp(cfpMsg);
            if (reply == null) return;

            if (cfpMsg.getConversationId() != null) reply.setConversationId(cfpMsg.getConversationId());
            fsm.sharedData.put("propose_msg", reply);
            fsm.getAgent().send(reply);
            transitionCode = (reply.getPerformative() == Performative.PROPOSE) ? 0 : 1;
            isDone = true;
        }
        @Override
        public boolean done() { return isDone; }
        @Override
        public int onEnd() { return transitionCode; }
    }

    private class WaitAcceptRejectBehavior extends SimpleBehavior {
        FIPAContractNetResponder fsm;
        int transitionCode;
        boolean isDone;

        @Override
        public void onStart() {
            super.onStart();
            this.transitionCode = -1;
            this.isDone = false;
        }

        WaitAcceptRejectBehavior(FIPAContractNetResponder fsm) { super(fsm.getAgent()); this.fsm = fsm; }
        @Override
        public void action() {
            ACLMessageTemplate template = (ACLMessageTemplate) fsm.sharedData.get("reply_template");
            ACLMessage reply = receive(template);
            if (reply == null) return;
            
            ACLMessage cfpMsg = (ACLMessage) fsm.sharedData.get("cfp");
            ACLMessage proposeMsg = (ACLMessage) fsm.sharedData.get("propose_msg");
            if (reply.getPerformative() == Performative.ACCEPT_PROPOSAL) {
                fsm.sharedData.put("accept_msg", reply);
                transitionCode = 0;
            } else {
                Boolean res = fsm.handleRejectProposal(cfpMsg, proposeMsg, reply);
                if (res == null || !res) return;
                transitionCode = 1;
            }
            isDone = true;
        }
        @Override
        public boolean done() { return isDone; }
        @Override
        public int onEnd() { return transitionCode; }
    }

    private class ExecuteTaskBehavior extends SimpleBehavior {
        FIPAContractNetResponder fsm;
        boolean isDone;

        @Override
        public void onStart() {
            super.onStart();
            this.isDone = false;
        }

        ExecuteTaskBehavior(FIPAContractNetResponder fsm) { super(fsm.getAgent()); this.fsm = fsm; }
        @Override
        public void action() {
            ACLMessage cfpMsg = (ACLMessage) fsm.sharedData.get("cfp");
            ACLMessage proposeMsg = (ACLMessage) fsm.sharedData.get("propose_msg");
            ACLMessage acceptMsg = (ACLMessage) fsm.sharedData.get("accept_msg");
            ACLMessage resultMsg = fsm.handleAcceptProposal(cfpMsg, proposeMsg, acceptMsg);
            if (resultMsg == null) return;

            if (cfpMsg.getConversationId() != null) resultMsg.setConversationId(cfpMsg.getConversationId());
            fsm.getAgent().send(resultMsg);
            isDone = true;
        }
        @Override
        public boolean done() { return isDone; }
    }
}
